src/CreditCardController.php

```diff
@@ -45,141 +68,182 @@ function validate(\Payment $payment, \PaymentMethod $method, $strict) {
-        '@type' => $err['type'],
-        '@param' => $err['param'],
-        '@message' => $err['message'],
-        '@pid'      => $payment->pid,
-        '@pmid'     => $payment->method->pmid,
-      ];
-      watchdog_exception('stripe_payment', $e, $message, $variables, WATCHDOG_ERROR);
-      $payment->setStatus(new \PaymentStatusItem(PAYMENT_STATUS_FAILED));
-      entity_save('payment', $payment);
-    }
-    catch (\Stripe\Error\Base $e) {
-      $payment->setStatus(new \PaymentStatusItem(PAYMENT_STATUS_FAILED));
-      entity_save('payment', $payment);
-
-      $message =
-        '@method payment method encountered an error while contacting ' .
-        'the stripe server. The status code "@status" and the error ' .
-        'message "@message". (pid: @pid, pmid: @pmid)';
-      $variables = array(
-        '@status'   => $e->getHttpStatus(),
-        '@message'  => $e->getMessage(),
-        '@pid'      => $payment->pid,
-        '@pmid'     => $payment->method->pmid,
-        '@method'   => $payment->method->title_specific,
-      );
-      watchdog_exception('stripe_payment', $e, $message, $variables, WATCHDOG_ERROR);
     }
+    entity_save('payment', $payment);
   }
 
+  /**
+   * Create a record for the {stripe_payment} table.
+   */
+  protected function createRecord($stripe, $plan_id = NULL) {
+    return [
+      // Subscription id (sub_) or payment intent id (pi_).
+      'stripe_id' => $stripe->id,
+      // "subscription" or "payment_intent".
+      'type'      => $stripe->object,
+      'plan_id'   => $plan_id,
+    ];
+  }
 
-  public function createCustomer($token, $description, $email) {
-    return \Stripe\Customer::create(array(
-        'card'        => $token,
-        'description' => $description,
-        'email'       => $email
-      ));
+  /**
+   * Create a new customer using the API.
+   */
+  public function createCustomer($intent, $context) {
+    return Customer::create([
+      'payment_method' => $intent->payment_method,
+      'invoice_settings' => [
+        'default_payment_method' => $intent->payment_method,
+      ],
+      // TODO: Use CreditCardForm::mappedFields()
+      'name' => $this->getName($context),
+      'email' => $context->value('email'),
+    ]);
   }
 
+  /**
+   * Calculate the total amount.
+   *
+   * @return int
+   *   Total amount converted to cents.
+   */
   public function getTotalAmount(\Payment $payment) {
-    // convert amount to cents. Integer value.
     return (int) ($payment->totalAmount(0) * 100);
   }
 
-  public function createCharge($customer, $payment) {
-    return \Stripe\Charge::create(array(
-        'customer' => $customer->id,
+  /**
+   * Create a new payment intent using the API.
+   */
+  public function createIntent($payment) {
+    // PaymentIntent: Make a payment immediately.
+    if ($this->filterRecurringLineItems($payment, FALSE)) {
+      return PaymentIntent::create([
         'amount'   => $this->getTotalAmount($payment),
```

Comment:
This gets the total of all line items. Even if some of them are recurring.

Reply:
That’s right, `getTotalAmout()` should only sum up one-off line-items!